sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/BodyMemory.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Azure.Core;
+using Microsoft.Azure.Amqp.Encoding;
+using Microsoft.Azure.Amqp.Framing;
+
+namespace Azure.Messaging.ServiceBus.Amqp
+{
+    internal sealed class BodyMemory : IEnumerable<ReadOnlyMemory<byte>>
+    {
+        private ArrayBufferWriter<byte> _writer;
+        private IList<ReadOnlyMemory<byte>> _segments;
+        private readonly IEnumerable<ReadOnlyMemory<byte>> _lazySegments;
+
+        private ReadOnlyMemory<byte> WrittenMemory
+        {
+            get
+            {
+                if (_lazySegments != null)
+                {
+                    foreach (var segment in _lazySegments)
+                    {
+                        Append(segment);
+                    }
+                }
+
+                return _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
+            }
+        }
+
+        // for the send path the combined memory for the body is not precomputed
+        private BodyMemory(IEnumerable<ReadOnlyMemory<byte>> dataSegments)
+        {
+            _lazySegments = dataSegments;
+        }
+
+        // for the receive path the combined memory is precomputed because we need to copy the underlying AMQP data anyway
+        private BodyMemory(IEnumerable<Data> dataSegments)
+        {
+            foreach (var segment in dataSegments)
+            {
+                Append(segment);
+            }
+        }
+
+        public static BodyMemory From(IEnumerable<ReadOnlyMemory<byte>> segments)
+        {
+            return segments as BodyMemory ?? new BodyMemory(segments);
```

Comment:
So if the user smuggled in `IEnumerable<ReadOnlyMemory<byte>>` over the lower level APIs it will be freshly wrapped and materialized when needed.